lib/layer.js

```diff
@@ -48,6 +48,31 @@ function Layer(path, options, fn) {
 }
 
 /**
+ * Handle any request for the layer.
+ *
+ * @api private
+ */
+
+Layer.prototype._handle_any = function _handle_any(/* ...args, next */) {
+  var self = this
+  var next = arguments[arguments.length - 1]
+  var handle = this.handle
+  var result
+  var hasNextBeenCalled = function () { return next._currentLayer === self }
+
+  try {
+    result = handle.apply(undefined, arguments)
+  } catch (err) {
+    return next(err)
+  }
+
+  if (result != null && typeof result.then === 'function') {
+    var onRejected = function (error) { if (!hasNextBeenCalled()) { next(error) } }
```

Comment:
Just a note, but we should probably use `next(error || new Error('something here')` as promises can be rejected with falsy values.